stakedb/ticketpool.go

```diff
@@ -225,18 +175,25 @@ func NewTicketPool(dataDir, dbSubDir string) (tp *TicketPool, err error) {
 	// Open ticket pool diffs database
 	badgerDbPath := filepath.Join(dataDir, dbSubDir)
 	opts := badger.DefaultOptions(badgerDbPath)
+	opts.MetricsEnabled = false
 	opts.Logger = &badgerLogger{log}
 	db, err := badger.Open(opts)
-	if err == badger.ErrTruncateNeeded {
-		log.Warnf("NewTicketPool badger db: %v", err)
-		// Try again with value log truncation enabled.
-		opts.Truncate = true
-		log.Warnf("Attempting to reopening ticket pool db with the Truncate option set...")
-		db, err = badger.Open(opts)
-	}
 	if err != nil {
-		return nil, fmt.Errorf("failed badger.Open: %v", err)
+		if !strings.Contains(err.Error(), "manifest has unsupported version") {
+			return nil, fmt.Errorf("failed to open badger DB: %w", err)
+		}
+
+		// Upgrade database.
+		if err = upgrade(badgerDbPath); err != nil {
```

Comment:
Only running `upgrade` on an error from `badger.Open` is not quite right.

Say there's a subsequent upgrade (but still using badger v3), it won't error, so we still need the `if ver < currentVersion` that you removed below.

Again, the db version integer we set has nothing to do with the badger version.

Reply:
> Say there's a subsequent upgrade (but still using badger v3), it won't error, so we still need the if ver < currentVersion that you removed below.

I removed that upgrade path since there was clearly no other upgrade but I have just added it in https://github.com/decred/dcrdata/pull/1948/commits/27a8d7453e3cddd9c2db4c088580a87a717d723a

Reply:
Thank you.  The upgrade system was incomplete without that.  When there's a future upgrade to the encoding, but with a version of badger that does not throw an error on Open, it the upgrade will be done via the version integer comparison.